tests/installation/npmTest.ts

```diff
@@ -0,0 +1,380 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { Expect } from '@playwright/test';
+import { test as _test, expect as _expect } from '@playwright/test';
+import fs from 'fs';
+import path from 'path';
+import type { Server } from 'http';
+import http from 'http';
+import https from 'https';
+import crypto from 'crypto';
+import type { SpawnOptions } from 'child_process';
+import debugLogger from 'debug';
+import { spawn } from 'child_process';
+
+const debug = debugLogger('itest');
+const debugExec = debugLogger('itest:exec');
+const debugExecStdout = debugLogger('itest:exec:stdout');
+const debugExecStderr = debugLogger('itest:exec:stderr');
+
+function spawnAsync(cmd: string, args: string[], options: SpawnOptions = {}): Promise<{stdout: string, stderr: string, code: number | null, error?: Error}> {
+  debugExec([cmd, ...args].join(' '));
+  const process = spawn(cmd, args, Object.assign({ windowsHide: true }, options));
+
+  return new Promise(resolve => {
+    let stdout = '';
+    let stderr = '';
+    if (process.stdout) {
+      process.stdout.on('data', data => {
+        debugExecStdout(data.toString());
+        stdout += data;
+      });
+    }
+    if (process.stderr) {
+      process.stderr.on('data', data => {
+        debugExecStderr(data.toString());
+        stderr += data;
+      });
+    }
+    process.on('close', code => resolve({ stdout, stderr, code }));
+    process.on('error', error => resolve({ stdout, stderr, code: 0, error }));
+  });
+}
+
+const kPublicNpmRegistry = 'https://registry.npmjs.org';
+const kContentTypeAbbreviatedMetadata = 'application/vnd.npm.install-v1+json';
+
+/**
+ * A minimal NPM Registry Server that can serve local packages, or proxy to the upstream registry.
+ * This is useful in test installation behavior of packages that aren't yet published. It's particularly helpful
+ * when your installation requires transitive dependencies that are also not yet published.
+ *
+ * See https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md for information on the offical APIs.
+ */
+
+_expect.extend({
+  toHaveDownloaded(received: any, browsers: ('chromium' | 'firefox' | 'webkit')[]) {
+    if (!(received instanceof ExecOutput))
+      throw new Error(`Expected ExecOutput instance.`);
+
+    const downloaded = new Set();
+    for (const [, browser] of received.combined().matchAll(/^.*(chromium|firefox|webkit) v\d+ downloaded.*$/img))
+      downloaded.add(browser);
+    try {
+      const expected = [...browsers];
+      browsers.sort();
+      const actual = [...downloaded];
+      actual.sort();
+      _expect(actual).toEqual(expected);
+    } catch (err) {
+      return {
+        message: () => `Browser download expectation failed:\n${err.toString()}`,
+        pass: false,
+      };
+    }
+
+    return {
+      pass: true,
+    };
+  }
+});
+
+interface CustomExpect extends Expect {
```

Comment:
We could extend the PlaywrightTest namespace see [here](https://playwright.dev/docs/test-advanced#add-custom-matchers-using-expectextend) instead of a custom interface, might be better and follow the official way.